File: grupo4/projC/src/Venda.c
#include "../include/Venda.h"

struct venda{
 Produto produto;
 Cliente cliente;
 float preco;
 int quant;
 char tipoCompra;
 int mes;
 int filial;
};

Venda inicializaV(){
 Venda v=(Venda)malloc(sizeof(struct venda));
 v->produto=inicializa_Prod();
 v->cliente=inicializa_Cliente();
 v->preco=0.0f;
 v->quant=0;
 v->mes=0;
 v->filial=0;
 return v;
}

gboolean validaV(Venda v,Cat_Produtos cp,Cat_Clientes cc){
 if (!((v->preco)>=0.0 && (v->preco)<=999.99))return FALSE;
 if (!((v->quant)>=1 && (v->quant)<=200)) return FALSE;
 if (!((v->tipoCompra=='N') ||(v->tipoCompra=='P'))) return FALSE;
 if (!((v->mes)>=1 && (v->mes)<=12)) return FALSE;
 if (!((v->filial)>=1 && (v->filial)<=3)) return FALSE;
 if (!validaProd(v->produto)) return FALSE;
 if (!validaCliente(v->cliente)) return FALSE;
 if (!existeProd(cp,v->produto)) return FALSE;
 if (!existeCliente(cc,v->cliente)) return FALSE;
 return TRUE;
}

Venda criaVendas(Venda v,char* p,char* c,float preco,int q,char tipoCompra ,int m,int f){
 v->produto = criaProduto(v->produto,p);
 v->cliente = criaCliente(v->cliente,c);
 v->preco = preco;
 v->quant = q;
 v->tipoCompra=tipoCompra;
 v->mes = m;
 v->filial = f;
 return v;
}

Produto getProdutoVenda(Venda v){
 return(v->produto);
}

float getPrecoVenda(Venda v){
 return (v->preco);
}

int getQntVenda(Venda v){
 return(v->quant);
}

char getTipo(Venda v) {
 return (v->tipoCompra);
}

Cliente getClienteVenda(Venda v) {
 return (v->cliente);
}

int getMes (Venda v){
 return (v->mes);
}

int getFilial (Venda v){
 return (v->filial);
}

void removeVenda (Venda v){
 removeCliente(v->cliente);
 removeProduto(v->produto);
 free(v);
}

File: grupo4/projJava/src/Models/Venda.java
package Models;

import Models.Catalogos.Produto;
import Common.Constantes;
import Models.Catalogos.CatClientes;
import Models.Catalogos.CatProds;
import Models.Catalogos.Cliente;

import java.util.Objects;

public class Venda {
    private Produto produto;
    private Cliente cliente;
    private float preco;
    private int quantidade;
    private char tipoCompra;
    private int mes;
    private int filial;


    public Venda() {
        this.produto = new Produto();
        this.cliente = new Cliente();
        this.preco = 0.0f;
        this.quantidade = 0;
        this.tipoCompra = 'N';
        this.mes = 0;
        this.filial = 0;
    }

    public Venda(Produto produto, Cliente cliente, float preco, int quantidade, char tipoCompra, int mes, int filial) {
        this.produto = new Produto(produto);
        this.cliente = new Cliente(cliente);
        this.preco = preco;
        this.quantidade = quantidade;
        this.tipoCompra = tipoCompra;
        this.mes = mes;
        this.filial = filial;
    }

    public Venda(Venda v){
        this.produto = v.getProduto();
        this.cliente = v.getCliente();
        this.preco = v.getPreco();
        this.quantidade = v.getQuantidade();
        this.tipoCompra = v.getTipoCompra();
        this.mes = v.getMes();
        this.filial = v.getFilial();
    }

    public Produto getProduto() {
        return produto.clone();
    }

    public void setProduto(Produto produto) {
        this.produto = new Produto(produto);
    }

    public Cliente getCliente() {
        return cliente.clone();
    }

    public void setCliente(Cliente cliente) {
        this.cliente = new Cliente(cliente);
    }

    public float getPreco() {
        return preco;
    }

    public void setPreco(float preco) {
        this.preco = preco;
    }

    public int getQuantidade() {
        return quantidade;
    }

    public void setQuantidade(int quantidade) {
        this.quantidade = quantidade;
    }

    public char getTipoCompra() {
        return tipoCompra;
    }

    public void setTipoCompra(char tipoCompra) {
        this.tipoCompra = tipoCompra;
    }

    public int getMes() {
        return mes;
    }

    public void setMes(int mes) {
        this.mes = mes;
    }

    public int getFilial() {
        return filial;
    }

    public void setFilial(int filial) {
        this.filial = filial;
    }

    public boolean equals(Object o){
        if (o == this) return true;
        if (o == null || o.getClass() != this.getClass()) return false;
        Venda a = (Venda) o;
        return this.produto.equals(a.getProduto())
                && this.cliente.equals(a.getCliente())
                && this.preco == (a.getPreco())
                && this.quantidade == (a.getQuantidade())
                && this.tipoCompra == (a.getTipoCompra())
                && this.mes == (a.getMes())
                && this.filial == (a.getFilial());
    }

    public String toString (){
        StringBuilder sb = new StringBuilder();
        sb.append("Código de model.Catalogos.Produto ").append(this.produto)
        .append("\nCódigo de model.Catalogos.Cliente:").append(this.cliente)
        .append("\nPreco:").append(this.preco)
        .append("\nQuantidade:").append(this.quantidade)
        .append("\nTipo de Compra:").append(this.tipoCompra)
        .append("\nMes:").append(this.mes)
        .append("\nFilial:").append(this.filial).append("\n");
        return sb.toString();
    }

    public Venda clone(){
        return new Venda(this);
    }

    public int hashCode() {
        return Objects.hash(produto);
    }

    /* Valida uma venda */
    public boolean validaV(CatProds cp, CatClientes cc){
       if (!((this.preco)>=0.0 && (this.preco)<=999.99))return false;
       if (!((this.quantidade)>=1 && (this.quantidade)<=200)) return false;
       if (!((this.tipoCompra=='N') ||(this.tipoCompra=='P'))) return false;
       if (!((this.mes)>=1 && (this.mes)<= Constantes.MESES)) return false;
       if (!((this.filial)>=1 && (this.filial)<=Constantes.FILIAIS)) return false;
       if (!this.produto.validaProd()) return false;
       if (!this.cliente.validaCliente()) return false;
       if (!cp.existeProduto(this.produto)) return false;
       if (!cc.existeCliente(this.cliente)) return false;

        return true;
    }


}
